attest/challenge.go

```diff
@@ -0,0 +1,144 @@
+package attest
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+)
+
+const (
+	ekBlobTag         = 0x000c
+	ekBlobActivateTag = 0x002b
+	ekTypeActivate    = 0x0001
+
+	algXOR = 0x0000000a
+
+	schemeESNone = 0x0001
+)
+
+type symKeyHeader struct {
+	Alg     uint32
+	Scheme  uint16
+	KeySize uint16
+}
+
+type activationBlobHeader struct {
+	Tag       uint16
+	KeyHeader symKeyHeader
+}
+
+func makeEmptyPCRInfo() []byte {
+	var b bytes.Buffer
+	binary.Write(&b, binary.BigEndian, uint16(3)) // SIZE_OF_SELECT
+	b.Write([]byte{0x00, 0x00, 0x00})             // empty bitfield for 3 PCRs
+	b.Write([]byte{0x01})                         // TPM_LOCALITY_SELECTION = TPM_LOC_ZERO
+	b.Write(bytes.Repeat([]byte{0}, sha1.Size))   // TPM_COMPOSITE_HASH
+	return b.Bytes()
+}
+
+func makeActivationBlob(symKey, aikpub []byte) (blob []byte, err error) {
+	aikHash := sha1.Sum(aikpub)
+
+	var out bytes.Buffer
+	if err := binary.Write(&out, binary.BigEndian, activationBlobHeader{
+		Tag: ekBlobActivateTag,
+		KeyHeader: symKeyHeader{
+			Alg:     algXOR,
+			Scheme:  schemeESNone,
+			KeySize: uint16(len(symKey)),
+		},
+	}); err != nil {
+		return nil, err
+	}
+
+	if _, err := out.Write(symKey); err != nil {
```

Comment:
nit: you don't need to check the Write error on a bytes.Buffer

```
out.Write(symKey)
out.Write(aikHash[:])
out.Write(makeEmptyPCRInfo())
```

Reply:
Done.